pq-crypto/bike/aes_ctr_prf.c

```diff
@@ -0,0 +1,109 @@
+/***************************************************************************
+* Additional implementation of "BIKE: Bit Flipping Key Encapsulation". 
+* Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Written by Nir Drucker and Shay Gueron
+* AWS Cryptographic Algorithms Group
+* (ndrucker@amazon.com, gueron@amazon.com)
+*
+* The license is detailed in the file LICENSE.txt, and applies to this file.
+* ***************************************************************************/
+
+#include <string.h>
+#include "aes_ctr_prf.h"
+#include "utilities.h"
+
+status_t init_aes_ctr_prf_state(OUT aes_ctr_prf_state_t *s,
+                                IN const uint32_t max_invokations,
+                                IN const seed_t *seed) 
+{
+        if (0 == max_invokations) {
+                return E_AES_CTR_PRF_INIT_FAIL;
+        }
+
+        // Set the key schedule (from seed).
+        // Make sure the size matches the AES256 key size
+        aes256_key_t key;
+        bike_static_assert(sizeof(seed->u) == sizeof(key.raw),
+                           seed_size_equals_ky_size);
+        memcpy(key.raw, seed->u.raw, sizeof(key.raw));
+
+        aes256_key_expansion(&s->ks, &key);
+
+        // Initialize buffer and counter
+        s->ctr.u.qw[0] = 0;
+        s->ctr.u.qw[1] = 0;
+        s->buffer.u.qw[0] = 0;
+        s->buffer.u.qw[1] = 0;
+
+        s->pos = AES256_BLOCK_SIZE;
+        s->rem_invokations = max_invokations;
+
+        SEDMSG("    Init aes_prf_ctr state:\n");
+        SEDMSG("      s.pos = %d\n", s->pos);
+        SEDMSG("      s.rem_invokations = %u\n", s->rem_invokations);
+        SEDMSG("      s.ctr = 0x");
+
+        secure_clean(key.raw, sizeof(key));
+
+        return SUCCESS;
+}
+
+_INLINE_ status_t perform_aes(OUT uint8_t *ct, IN OUT aes_ctr_prf_state_t *s) {
+        // Ensure that the CTR is big enough
+        bike_static_assert(((sizeof(s->ctr.u.qw[0]) == 8) && (BIT(33) >= MAX_AES_INVOKATION)),
+                           ctr_size_is_too_small);
+
+        if (0 == s->rem_invokations) {
+            return E_AES_OVER_USED;
+        }
+
+        aes256_enc(ct, s->ctr.u.bytes, &s->ks);
+
+        s->ctr.u.qw[0]++;
+        s->rem_invokations--;
+
+        return SUCCESS;
+}
+
+status_t aes_ctr_prf(OUT uint8_t *a,
+                     IN OUT aes_ctr_prf_state_t *s,
+                     IN const uint32_t len)
+{
+    status_t res = SUCCESS;
```

Comment:
There's no way for this function to ever fail. 

GUARD is defined as `#define GUARD(func, stat, label) {stat = func; if(stat != SUCCESS) {goto label;}}`

and all the GUARD will `goto EXIT;` which just returns `SUCCESS` even on error

Reply:
Ah, nevermind, this will fail if the underlying function fails. The `GUARD` macro is overwriting `res` through the `stat` Macro variable.